pipeline/trusted-resources/pkg/trustedtask/trustedtaskrun.go

```diff
@@ -101,125 +92,58 @@ func (tr *TrustedTaskRun) VerifyTask(
-	} else {
-		cosignPublicKeypath := secretPath
-		// Overwrite the path if set in configmap.
-		cm, err := k8sclient.CoreV1().ConfigMaps(system.Namespace()).Get(ctx, signingConfigMap, metav1.GetOptions{})
-		if err != nil {
-			return nil, err
-		}
-		if cm.Data["signing-secret-path"] != "" {
-			cosignPublicKeypath = cm.Data["signing-secret-path"]
-		}
-		return cosignsignature.LoadPublicKey(ctx, cosignPublicKeypath)
-	}
-}
+	delete(cp.ObjectMeta.Annotations, SignatureAnnotation)
 
-func VerifyInterface(
-	ctx context.Context,
-	taskspec interface{},
-	verifier signature.Verifier,
-	signature []byte,
-) (errs *apis.FieldError) {
-	ts, err := json.Marshal(taskspec)
+	verifier, err := verifier(ctx, cp.ObjectMeta.Annotations, k8sclient)
 	if err != nil {
-		return apis.ErrGeneric(err.Error(), "taskSpec")
+		return apis.ErrGeneric(err.Error(), "metadata")
 	}
 
-	h := sha256.New()
-	h.Write(ts)
+	if err := VerifyInterface(ctx, cp, verifier, signature); err != nil {
+		return apis.ErrGeneric(err.Error(), "taskrun")
+	}
 
-	if err := verifier.VerifySignature(bytes.NewReader(signature), bytes.NewReader(h.Sum(nil))); err != nil {
-		return apis.ErrGeneric(err.Error(), "taskSpec")
+	if tr.Spec.TaskRef != nil {
+		// TODO: add task verification in next PR
 	}
 
 	return nil
 }
 
-func VerifyTaskOCIBundle(
-	ctx context.Context,
-	bundle string,
-	verifier signature.Verifier,
-	signature []byte,
-	k8sclient kubernetes.Interface,
-) (errs *apis.FieldError) {
+// copyTaskRun will exclude system populated fields and return copied taskrun
+func copyTaskRun(in *TrustedTaskRun) (v1beta1.TaskRun, error) {
+	c := v1beta1.TaskRun{}
+	c.TypeMeta = in.TypeMeta
+	c.SetName(in.Name)
+	c.SetGenerateName(in.GenerateName)
+	c.SetNamespace(in.Namespace)
 
-	serviceAccountName := os.Getenv("WEBHOOK_SERVICEACCOUNT_NAME")
-	if serviceAccountName == "" {
-		serviceAccountName = "tekton-verify-task-webhook"
-	}
-	kc, err := k8schain.New(ctx, k8sclient, k8schain.Options{
-		Namespace:          system.Namespace(),
-		ServiceAccountName: serviceAccountName,
-	})
-	if err != nil {
-		return apis.ErrGeneric(err.Error()).ViaKey(bundle)
+	c.Labels = make(map[string]string)
+	for k, v := range in.Labels {
+		c.Labels[k] = v
 	}
+	delete(c.ObjectMeta.Labels, "app.kubernetes.io/managed-by")
 
-	timeoutCtx, cancel := context.WithTimeout(ctx, time.Second*60)
-	defer cancel()
-
-	digest, err := Digest(ctx, bundle, remote.WithAuthFromKeychain(kc), remote.WithContext(timeoutCtx))
-	if err != nil {
-		return apis.ErrGeneric(err.Error()).ViaKey(bundle)
+	c.Annotations = make(map[string]string)
+	for k, v := range in.Annotations {
+		c.Annotations[k] = v
 	}
+	delete(c.ObjectMeta.Annotations, "kubectl.kubernetes.io/last-applied-configuration")
 
-	if err := verifier.VerifySignature(bytes.NewReader(signature), bytes.NewReader([]byte(digest.String()))); err != nil {
-		return apis.ErrGeneric(err.Error()).ViaKey(bundle)
+	c.Spec = in.Spec
+	if c.Spec.Timeout != nil {
+		c.Spec.Timeout = nil
 	}
```

Comment:
I'm very hesitant about doing this - we should generally respect fields set in the user spec.

Reply:
Oh yes. I thought this is generated by system. This param has default value 1 hour, so if user doesn't specify this field, we will get value at the webook. Not sure how to deal with this situation. Maybe we should force user to set this?

Reply:
I think one solution is to add the default value in our signing cmd, if this is not defined by user then we will add the default value and document this.

Reply:
🤔  We should think more about how we should handle mutating webhooks - these are just values set by the pipelines admission controller at runtime, but there could be others that mutate other values.

Reply:
- ah yes, the defaulting mutating webhook will call `SetDefaults()` and add some defaults data. So we can also call this function to add the defaults values. But this is not feasible since you mentioned there may be other mutating wh that can change the data at admission time.

- Another way is try to validate before the mutating?